scripts/build/build/factory.py

```diff
@@ -0,0 +1,127 @@
+import os
+
+from typing import Set
+
+from builders.builder import Builder
+from builders.linux import LinuxBuilder
+from builders.qpg import QpgBuilder
+from builders.esp32 import Esp32Builder, Esp32Board
+from builders.efr32 import Efr32Builder, Efr32App, Efr32Board
+
+from .targets import Application, Board, Platform
+
+
+class Matcher():
+  """Figures out if a proper builder can be created for a platform/board/app combination."""
+
+  def __init__(self, builder_class):
+    self.builder_class = builder_class
+    self.app_arguments = {}
+    self.board_arguments = {}
+
+  def AcceptApplication(self, __app_key: Application, **kargs):
+    self.app_arguments[__app_key] = kargs
+
+  def AcceptBoard(self, __board_key: Board, **kargs):
+    self.board_arguments[__board_key] = kargs
+
+  def Create(self, __board_key: Board, __app_key: Application, repo_path: str,
+             **kargs):
+    """Creates a new builder for the given board/app. """
+    if not __board_key in self.board_arguments:
+      return None
+
+    if not __app_key in self.app_arguments:
+      return None
+
+    kargs.update(self.board_arguments[__board_key])
+    kargs.update(self.app_arguments[__app_key])
+
+    return self.builder_class(repo_path, **kargs)
+
+
+# Builds a list of acceptable application/board combination for every platform
+_MATCHERS = {
+    Platform.LINUX: Matcher(LinuxBuilder),
+    Platform.ESP32: Matcher(Esp32Builder),
+    Platform.QPG: Matcher(QpgBuilder),
+    Platform.EFR32: Matcher(Efr32Builder),
+}
+
+# Matrix of what can be compiled using and what build options are required
+# by such compilation
+_MATCHERS[Platform.LINUX].AcceptApplication(Application.ALL_CLUSTERS)
+_MATCHERS[Platform.LINUX].AcceptBoard(Board.NATIVE)
+
+_MATCHERS[Platform.ESP32].AcceptApplication(Application.ALL_CLUSTERS)
+_MATCHERS[Platform.ESP32].AcceptBoard(Board.DEVKITC, board=Esp32Board.DevKitC)
+_MATCHERS[Platform.ESP32].AcceptBoard(Board.M5STACK, board=Esp32Board.M5Stack)
+
+_MATCHERS[Platform.QPG].AcceptApplication(Application.LOCK)
+_MATCHERS[Platform.QPG].AcceptBoard(Board.QPG6100)
+
+_MATCHERS[Platform.EFR32].AcceptApplication(Application.LOCK)
+_MATCHERS[Platform.EFR32].AcceptBoard(Board.BRD4161A, board=Efr32Board.BRD4161A)
+_MATCHERS[Platform.EFR32].AcceptApplication(
+    Application.LIGHT, app=Efr32App.LIGHT)
+_MATCHERS[Platform.EFR32].AcceptApplication(Application.LOCK, app=Efr32App.LOCK)
```

Comment:
Why the "app" arg here but not for QPG or the EFR32 bit above?

Reply:
I created every builder independently and currently the QPG one compiles just one thing.
As this evolves I expect every 'Accept' will do so with an argument.

Agree the is inconsistency: EFR32 supports only one board, yet it has a board argument. The reason I did that was because our existing example script has a board argument as well so I tried to pass it in.

I expect the scripts to evolve over time for sure.

Reply:
I was specifically asking about the inconsistency between:
```
_MATCHERS[Platform.EFR32].AcceptApplication(Application.LOCK)
```
and
```
_MATCHERS[Platform.EFR32].AcceptApplication(Application.LOCK, app=Efr32App.LOCK)
```

Should one of those lines not be there?  Or do they do different things?

Reply:
oops - I missed that one. I believe this was some copy and paste bug not noticed during running because 'last set wins'.

Updated. Thank you for checking.

Reply:
> Agree the is inconsistency: EFR32 supports only one board, yet it has a board argument. The reason I did that was because our existing example script has a board argument as well so I tried to pass it in.

Untrue, EFR32 supports 7 boards.